lib/stinger_core/inc/stinger_batch_insert.h

```diff
@@ -0,0 +1,571 @@
+            }
+
+            // Combine all ranges into shared list
+            OMP("omp critical")
+            update_ranges.insert(update_ranges.end(), local_ranges.begin(), local_ranges.end());
+        }
+
+
+        LOG_V_A("Entering parallel update loop: %ld updates for %ld vertices.",
+            std::distance(updates_begin, updates_end), unique_sources.size()-1);
+        OMP("omp parallel for schedule(dynamic)")
+        for (range_iterator range = update_ranges.begin(); range < update_ranges.end(); ++range)
+        {
+            // Get this thread's range of updates
+            iterator begin = range->first;
+            iterator end = range->second;
+            // Each range guaranteed to have same edge type and source vertex
+            int64_t type = adapter::get_type(*begin);
+            int64_t source = use_source::get(*begin);
+
+            LOG_D_A("Thread %d processing %ld updates (%ld -> *)", omp_get_thread_num(), std::distance(begin, end), source);
+            update_directed_edges_for_vertex<direction, use_dest>(G, source, type, begin, end, operation);
+        }
+    }
+
+    static bool
+    source_less_than_destination(const update &x){
+        return adapter::get_source(x) < adapter::get_dest(x);
+    }
+
+    // We use the result field to keep track of which updates have been performed
+    // Between IN and OUT iterations, we clear it out to make sure all the updates are performed again
+    static void
+    clear_results(iterator begin, iterator end)
+    {
+        OMP("omp parallel for")
+        for (iterator u = begin; u < end; ++u) {
+            int64_t result = adapter::get_result(*u);
+            switch (result){
+                case EDGE_ADDED:
+                case EDGE_UPDATED:
+                case EDGE_NOT_ADDED:
+                    // Reset return code so we process this update next iteration
+                    adapter::set_result(*u, PENDING);
+                    break;
+                case -1:
+                case 1:
+                    // Caller must have set the return code, leave it be
+                    break;
+                case PENDING:
+                default:
+                    // We didn't finish all the updates, or invalid codes were set
+                    assert(0);
+
+            }
+        }
+    }
+
+    static void
+    remap_results(iterator begin, iterator end)
+    {
+        OMP("omp parallel for")
+        for (iterator u = begin; u < end; ++u) {
+            int64_t result = adapter::get_result(*u);
+            switch (result){
+                case EDGE_ADDED:
+                case EDGE_UPDATED:
+                case EDGE_NOT_ADDED:
+                    // Subtract 10 to get back to return code caller expects
+                    adapter::set_result(*u, result - 10);
+                    break;
+                case -1:
+                case 1:
+                    // Caller must have set the return code, leave it be
```

Comment:
Unfortunately the calling code uses magic numbers too, but I suppose I could add these to the enum.